pysteps/nowcasts/steps.py

```diff
@@ -33,6 +33,873 @@
+            for _ in range(self.n_ens_members)
+        ]
+
+        # Initialize random generators if noise_method is provided
+        if self.noise_method is not None:
+            self.random_generator_precip = []
+            self.random_generator_motion = []
+
+            for _ in range(self.n_ens_members):
+                # Create random state for precipitation noise generator
+                rs = np.random.RandomState(self.seed)
+                self.random_generator_precip.append(rs)
+                self.seed = rs.randint(0, high=int(1e9))  # Update seed after generating
+
+                # Create random state for motion perturbations generator
+                rs = np.random.RandomState(self.seed)
+                self.random_generator_motion.append(rs)
+                self.seed = rs.randint(0, high=int(1e9))  # Update seed after generating
+        else:
+            self.random_generator_precip = None
+            self.random_generator_motion = None
+        print("AR model and noise applied to precipitation cascades.")
+
+    def _initialize_velocity_perturbations(self):
+        """
+        Initialize the velocity perturbators for each ensemble member if the velocity
+        perturbation method is specified.
+        """
+        if self.velocity_perturbation_method is not None:
+            init_vel_noise, generate_vel_noise = noise.get_method(
+                self.velocity_perturbation_method
+            )
+
+            self.velocity_perturbations = []
+            for j in range(self.n_ens_members):
+                kwargs = {
+                    "randstate": self.random_generator_motion[j],
+                    "p_par": self.velocity_perturbation_kwargs.get(
+                        "p_par", self.velocity_perturbation_parallel
+                    ),
+                    "p_perp": self.velocity_perturbation_kwargs.get(
+                        "p_perp", self.velocity_perturbation_perpendicular
+                    ),
+                }
+                vp = init_vel_noise(
+                    self.velocity, 1.0 / self.kmperpixel, self.timestep, **kwargs
+                )
+                self.velocity_perturbations.append(
+                    lambda t, vp=vp: generate_vel_noise(vp, t * self.timestep)
+                )
+        else:
+            self.velocity_perturbations = None
+        print("Velocity perturbations initialized successfully.")
+
+    def _initialize_precipitation_mask(self):
+        """
+        Initialize the precipitation mask and handle different mask methods (sprog, incremental).
+        """
+        self.precip_forecast = [[] for _ in range(self.n_ens_members)]
+
+        if self.probmatching_method == "mean":
+            self.mu_0 = np.mean(
+                self.precip[-1, :, :][self.precip[-1, :, :] >= self.precip_threshold]
+            )
+        else:
+            self.mu_0 = None
+
+        self.precip_mask = None
```

Comment:
Maybe, if I read the code correctly, you could create an empty `state` and `params` object (dataclass) in the `__init__` or at the start of `compute_forecast`. Then these objects can be passed along (or stored in self) to the methods that fill this state/params. This would group the currently existing variables a bit nicer together maybe?

Reply:
I indeed missed this one... Currently I added it to the init. Next step is indeed to add the two suggested classes to remove the number of class variables